modules/ocf_www/files/build-vhosts

```diff
@@ -14,190 +11,350 @@ from pathlib import Path
 
 import jinja2
 from ocflib.account.search import user_is_sorried
-from ocflib.account.utils import web_dir
 from ocflib.misc.mail import email_for_user
+from ocflib.vhost.application import get_app_vhosts
 from ocflib.vhost.web import get_vhosts
 
+
+APACHE_SITE_CONFIG = '/etc/apache2/ocf-vhost.conf'
+NGINX_SITE_CONFIG = '/etc/nginx/sites-enabled/virtual'
+
+LETS_ENCRYPT_SSL = Path('/services/http/ssl')
+SSL_BUNDLE = Path('/etc/ssl/apphost')
+
+APP_DIR = Path('/srv/apps')
+WEB_DIR = Path('/services/http/users')
+
+
 jinja_env = jinja2.Environment(
     loader=jinja2.FileSystemLoader((
-        os.path.abspath(os.path.dirname(__file__)),
-        '/opt/share/',
+        os.path.abspath('/opt/share'),
     )),
 )
 
 
-LETS_ENCRYPT_SSL = Path('/services/http/ssl')
+class SSL(namedtuple('SSL', ('fqdn',))):
+    """Tells where to find SSL files for a given domain."""
+    @property
+    def is_valid(self):
+        return os.path.exists(self.cert)
+
+    @property
+    def key(self):
+        return '/etc/ssl/lets-encrypt/le-vhost.key'
+
+    @property
+    def cert(self):
+        return str(LETS_ENCRYPT_SSL / (self.fqdn + '.crt'))
+
+    @property
+    def chain(self):
+        return '/etc/ssl/certs/lets-encrypt.crt'
+
+    @property
+    def bundle(self):
+        return str(SSL_BUNDLE / (self.fqdn + '.bundle'))
 
 
 class VirtualHost(namedtuple('VirtualHost', (
-        'fqdn',
-        'user',
-        'relative_docroot',
-        'redirect_dest',
-        'ssl_cert',
-        'ssl_key',
-        'ssl_chain',
-        'use_hsts',
-        'disabled',
+    'fqdn',
+    'user',
+    'comment',
+    'ssl',
+    'socket_name',
+    'relative_docroot',
+    'redirect_dest',
 ))):
-    """A single <VirtualHost>.
+    """A single virtual host directive in one of the webserver configs.
 
-    The VirtualHost has a one-to-one mapping to <VirtualHost> directives and is
-    rendered by one invocation of the template.
+    A VirtualHost has a one-to-one mapping to <VirtualHost> (apache) or
+    server (nginx) directives and is rendered by one invocation of the
+    template. This means that some vhost lines might emit multiple
+    VirtualHost instances (redirect domains, etc.).
```

Comment:
Yeah, this is probably better done somehow with polymorphism. I did have a similar class hierarchy earlier but it had large, repetitive constructors; I bet I can improve on that though. Something like this would probably also work but I think it might result in either a bunch of boilerplate functions to delegate to the `vhost` object or I would have to stick `.vhost` accessors in a bunch of places in the template files.